pallets/phala-world/src/lib.rs

```diff
@@ -0,0 +1,814 @@
+		#[transactional]
+		pub fn buy_rare_egg(
+			origin: OriginFor<T>,
+			egg_type: EggType,
+			race: RaceType,
+			career: CareerType,
+			metadata: BoundedVec<u8, T::StringLimit>,
+		) -> DispatchResult {
+			ensure!(CanPurchaseRareEggs::<T>::get(), Error::<T>::RareEggPurchaseNotAvailable);
+			let sender = ensure_signed(origin.clone())?;
+			let overlord = <Overlord<T>>::get();
+			match overlord {
+				None => Err(Error::<T>::OverlordNotSet.into()),
+				Some(overlord) => {
+					// Get Egg Price based on EggType
+					let egg_price = match egg_type {
+						EggType::Founder => {
+							T::FounderEggPrice::get()
+						},
+						EggType::Legendary => {
+							T::LegendaryEggPrice::get()
+						},
+						_ => T::NormalEggPrice::get(),
+					};
+					ensure!(egg_price != T::NormalEggPrice::get(), Error::<T>::InvalidPurchase);
+					let nft_id = pallet_rmrk_core::NextNftId::<T>::get(EGGS_COLLECTION_ID);
+					// Transfer the amount for the rare Egg NFT then mint the egg
+					<T as pallet::Config>::Currency::transfer(
+						&sender,
+						&overlord,
+						egg_price,
+						ExistenceRequirement::KeepAlive
+					)?;
+					// Mint Egg and transfer Egg to new owner
+					pallet_rmrk_core::Pallet::<T>::mint_nft(
+						Origin::<T>::Signed(overlord.clone()).into(),
+						sender.clone(),
+						EGGS_COLLECTION_ID,
+						None,
+						None,
+						metadata,
+					)?;
+					// Add EggInfo to storage
+					let egg = EggInfo {
+						egg_type,
+						race,
+						career,
+						start_hatching: 0,
+						hatching_duration: 0,
+					};
+					Eggs::<T>::insert(EGGS_COLLECTION_ID, nft_id, egg);
+
+					Self::deposit_event(Event::RareEggPurchased {
+						collection_id: EGGS_COLLECTION_ID,
+						nft_id,
+						owner: sender,
+					});
+
+					Ok(())
+				}
+			}
+		}
+
+		/// Users can pre-order an egg. This will enable users that are whitelisted to be
+		/// added to the queue of users that can claim eggs. Those that come after the whitelist
+		/// pre-sale will be able to win the chance to acquire an egg based on their choice of
+		/// race and career as they will have a limited quantity.
+		///
+		/// Parameters:
+		/// - origin: The origin of the extrinsic preordering the egg
+		/// - race: The race that the user has chosen (limited # of races)
+		/// - career: The career that the user has chosen (limited # of careers)
+		#[pallet::weight(10_000 + T::DbWeight::get().reads_writes(1,1))]
+		#[transactional]
+		pub fn preorder_egg(
+			origin: OriginFor<T>,
+			race: RaceType,
+			career: CareerType,
+		) -> DispatchResult {
+			ensure!(CanPreorderEggs::<T>::get(), Error::<T>::PreorderEggNotAvailable);
+			let sender = ensure_signed(origin)?;
+
+			// Reserve currency for the preorder at the Normal egg price
+			<T as pallet::Config>::Currency::reserve(&sender, T::NormalEggPrice::get())?;
+
+			let preorder = PreorderInfo {
+				owner: sender.clone(),
+				race,
+				career,
+			};
+			<Preorders<T>>::get().push(preorder.into());
```

Comment:
It implies you:

1. Read the entire list from the storage
2. Push a preorder to the list
3. Discard the list (as a temporary variable)

Instead, you can call `Preorders::<T>::append(preorder)`. It will make sure a new element will be appended to the list in the storage.

However, by doing so, essentially we have created a growing size `Vec<u8>` as a single entry in the storage. Whenever you read or write to that storage item, we operate on the full vector. So I doubt it can scale to more than a few hundreds items. We may consider to use a `StorageMap` where each read / write only operates on a single entry of the storage.